### PR TITLE
Making initialization state clearer

### PR DESCRIPTION
**Pull request by leandroBorgesFerreira**
### 🎯 Goal

Allow users to understand the initialization process of the SDK and decouple it from the connection of the SDK.

### 🛠 Implementation details

`ClientState.initialized` now returns an enum with stages instead of a boolean that set to true once the SDK is connected. 

This is a breaking change. Nevertheless, the current implementation is a bit misleading, because it is not clear how the initialization process is different from the connection process, which can cause confusion when waiting or asserting the initialization of the SDK.

### 🧪 Testing
Check the initialization of the SDK.

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- ~[ ] New code is covered by unit tests~
- ~[ ] Comparison screenshots added for visual changes~
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF
giphy
_The SDK is setting up_

### CODE REVIEW

**Review by JcMinarro — CHANGES_REQUESTED** [review]
I have left some comments.
On the other hand, could we have some tests that verify the `initializationState` is properly updated when the user is connected/disconnected?
We already have a `ConnectUserTests` class that cover connection process, maybe we can reuse those tests

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 1]
```diff
@@ -484,6 +489,7 @@ internal constructor(
         tokenProvider: TokenProvider,
         timeoutMilliseconds: Long? = null,
     ): Call<ConnectionData> {
+        clientState.toMutableState()?.setInitializionState(InitializationState.RUNNING)
```
[the remark this answers is not in the document]
Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:492]: It prevents users to try to connect a second time before the initialization is complete. If users have a use case that they can try to reconnect for a reason, they can check that the initialisation was already requested, but it is not completed yet. 

If we can have NotConnected/Connected it won't be possible to check the case that the initialisation was requested, but it is not completed yet.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/models/InitializationState.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * The initialization of the SDK was not requested. Use ChatClient.connectUser to start the
+     * initialization process.
+     */
+    NOT_INITIALIZED,
+
+    /**
+     * An error happened during the initialization process.
+     */
+    ERROR
```
[the remark this answers is not in the document]
Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/models/InitializationState.kt:45]: Then we could keep the boolean. Although the `ERROR` can be checked using the Result of the `connectUser`, but removing RUNNING would remove the possibility to check is the connection was already requested, but not completed yet.

File: stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/global/GlobalState.kt [hunk 3]
```diff
@@ -59,7 +60,7 @@ public interface GlobalState {
             ]
         )
     )
-    public val initialized: StateFlow<Boolean>
+    public val initialized: Flow<Boolean>
```
Comment (adasiewiczr) [inline stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/global/GlobalState.kt:63]: Why are we converting this to `Flow`? 🤔 I would rather keep it as a `StateFlow` and use `.stateIn()` in `GlobalMutablEstate`

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/setup/state/ClientState.kt [hunk 4]
```diff
@@ -32,9 +33,9 @@ public interface ClientState {
     public val user: StateFlow<User?>
 
     /**
-     * If the client connection has been initialized.
+     * The state of the initialization process of the SDK.
      */
-    public val initialized: StateFlow<Boolean>
+    public val initialized: StateFlow<InitializationState>
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/setup/state/ClientState.kt:38]: Can we implement it in a non-breaking way? Maybe we can add additional property `public val initializationState: StateFlow<InitializationState>` and deprecate the old one?

File: CHANGELOG.md [hunk 5]
```diff
@@ -19,6 +19,7 @@
 ### ⬆️ Improved
 - Offline data is clear after the user is disconnect by calling `ChatClient.disconnect(true)`. [#3917](https://github.com/GetStream/stream-chat-android/pull/3917)
 - Adding logs to understand more about unrecoverable errors in socket connection. [#3946](https://github.com/GetStream/stream-chat-android/pull/3946)
+- Refactor in the initialization state. Now you can check when the current state of the intialization progress and check if an error happened. [#3962](https://github.com/GetStream/stream-chat-android/pull/3962)
```
Comment (adasiewiczr) [inline CHANGELOG.md:22]: Can we add more info here? Let's mention which property was added/changed

Comment (JcMinarro) [reply CHANGELOG.md:22]: This text need to be updated, There is not "an error" state now

File: CHANGELOG.md [hunk 6]
```diff
@@ -19,6 +19,7 @@
 ### ⬆️ Improved
 - Offline data is clear after the user is disconnect by calling `ChatClient.disconnect(true)`. [#3917](https://github.com/GetStream/stream-chat-android/pull/3917)
 - Adding logs to understand more about unrecoverable errors in socket connection. [#3946](https://github.com/GetStream/stream-chat-android/pull/3946)
+- Added in the ChatClient.initializationState. Now you can check when the current state of the intialization progress. [#3962](https://github.com/GetStream/stream-chat-android/pull/3962)
```
Comment (JcMinarro) [inline CHANGELOG.md:22]:
```suggestion
- Added in the ClientState.initializationState. Now you can check when the current state of the intialization progress. [#3962](https://github.com/GetStream/stream-chat-android/pull/3962)
```

File: stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/global/GlobalState.kt (hunk 3 above)
Comment (adasiewiczr) [reply stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/global/GlobalState.kt:65]: Can we revert this change?

Comment (leandroBorgesFerreira) [reply stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/global/GlobalState.kt:65]: Yep, no problem

File: CHANGELOG.md [hunk 7]
```diff
@@ -19,6 +19,7 @@
 ### ⬆️ Improved
 - Offline data is clear after the user is disconnect by calling `ChatClient.disconnect(true)`. [#3917](https://github.com/GetStream/stream-chat-android/pull/3917)
 - Adding logs to understand more about unrecoverable errors in socket connection. [#3946](https://github.com/GetStream/stream-chat-android/pull/3946)
+- Added in the ClientState.initializationState. Now you can check when the current state of the intialization progress. [#3962](https://github.com/GetStream/stream-chat-android/pull/3962)
```
Comment (adasiewiczr) [inline CHANGELOG.md:22]:
```suggestion
- Added the ClientState.initializationState. Now you can check when the current state of the initialization progress. [#3962](https://github.com/GetStream/stream-chat-android/pull/3962)
```

**Review by adasiewiczr — APPROVED** [review]
LGTM

**Review by JcMinarro — APPROVED** [review]
LGTM

**Pull request merged**